Recompact fail with "failed recompaction: No such file or directory"

In rare cases, on our builds we got this error during `recompact` step:

```
'C:/ProgramData/chocolatey/bin/ninja.exe' '-C' '<SOME_PATH>' '-t' 'recompact'
    failed with:
      ninja: warning: bad deps log signature or version; starting over
    ninja: error: failed recompaction: No such file or directory  
```

Trying to trace back why this is failing, we see that:

- `DepsLog::Load` delete `ninja_deps` and return `LOAD_SUCCESS` if version mismatch or headers are corrupted.
- `OpenDepsLog(/*recompact_only=*/ true)` doesn't stop because on`LOAD_SUCCESS`
- `DepsLog::Recompact` fail because of the missing `ninja_deps` file.

In our case we need to find why the file loading fail (corrupted file ?) but I wonder if this is intentionnal to fail in this case ? 

- if no, returning a `LOAD_NOT_FOUND` instead of `LOAD_SUCCESS` will skip the recompaction since the file was deleted.
- if yes, a more explicit error message could help

thanks,

We started seeing this problem on our build farm after we've upgraded from 1.8 to 1.10.
After some investigation it turned out that the problem happens if the following steps are executed:
1. setup minimal cmake project
2. install ninja 1.8
3. build the project with it using `cmake <path> -G Ninja; ninja`
4. install ninja 1.10
5. run `cmake .` in the build folder.

These steps reliably reproduce the problem.
The problem happens even in 1.11 or master.

Here is the full log of commands:
```
 ~/projects/temp/cmake_minimal/build 0 > ninja-build --version
1.8.2
 ~/projects/temp/cmake_minimal/build 0 > cmake .. -G Ninja
-- The C compiler identification is GNU 4.8.5
-- The CXX compiler identification is GNU 4.8.5
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Configuring done
-- Generating done
-- Build files have been written to: /home/obf7/projects/temp/cmake_minimal/build
 ~/projects/temp/cmake_minimal/build 0 > ninja-build 
[2/2] Linking CXX executable cmake_minimal
 ~/projects/temp/cmake_minimal/build 0 > ninja-build --version
1.11.0
 ~/projects/temp/cmake_minimal/build 0 > cmake .
-- Configuring done
-- Generating done
CMake Error:
  Running

   '/usr/bin/ninja-build' '-C' '/home/obf7/projects/temp/cmake_minimal/build' '-t' 'recompact'

  failed with:

   ninja: warning: bad deps log signature or version; starting over

  ninja: error: failed recompaction: No such file or directory



CMake Generate step failed.  Build files cannot be regenerated correctly.
 ~/projects/temp/cmake_minimal/build 1 > 
```

OS: CentOS 7
CMake: 3.22.2

--- a/misc/output_test.py
+++ b/misc/output_test.py
@@ -127,6 +127,29 @@ build a: cat
         self.assertEqual(run('', flags='-t recompact'), '')
         self.assertEqual(run('', flags='-t restat'), '')
 
+    def test_issue_2048(self):
+        with tempfile.TemporaryDirectory() as d:
+            with open(os.path.join(d, 'build.ninja'), 'w'):
+                pass
+
+            with open(os.path.join(d, '.ninja_log'), 'w') as f:
+                f.write('# ninja log v4\n')
+
+            try:
+                output = subprocess.check_output([NINJA_PATH, '-t', 'recompact'],
+                                                 cwd=d,
+                                                 env=default_env,
+                                                 stderr=subprocess.STDOUT,
+                                                 text=True
+                                                 )
+
+                self.assertEqual(
+                    output.strip(),
+                    "ninja: warning: build log version is too old; starting over"
+                )
+            except subprocess.CalledProcessError as err:
+                self.fail("non-zero exit code with: " + err.output)
+
     def test_status(self):
         self.assertEqual(run(''), 'ninja: no work to do.\n')
         self.assertEqual(run('', pipe=True), 'ninja: no work to do.\n')

--- a/src/build_log.cc
+++ b/src/build_log.cc
@@ -291,9 +291,9 @@ LoadStatus BuildLog::Load(const string& path, string* err) {
       if (invalid_log_version) {
         fclose(file);
         unlink(path.c_str());
-        // Don't report this as a failure.  An empty build log will cause
+        // Don't report this as a failure. A missing build log will cause
         // us to rebuild the outputs anyway.
-        return LOAD_SUCCESS;
+        return LOAD_NOT_FOUND;
       }
     }
 

